File: problems/05/valid_palindrome.c
#include <criterion/criterion.h>
#include <stdint.h>

int valid_palindrome(const char *s, size_t len)
{
  const char *end = s + len - 1;
  while (s < end)
  {
    if (*s != *end)
    {
      return 0;
    }
    s++;
    end--;
  }

  return 1;
}

Test(ValidPalindrome, ValidPalindromeTests)
{
  cr_expect(valid_palindrome("racecar", 7) == 1);
  cr_expect(valid_palindrome("panama", 6) == 0);
  cr_expect(valid_palindrome("amanaplanacanalpanama", 21) == 1);
}

File: problems/05/valid_palindrome.exs
defmodule ValidPalindrome do
  require Integer

  def valid?(s) do
    valid_palindrome(s)
  end

  defp valid_palindrome(s) do
    mid = Integer.floor_div(String.length(s), 2)

    front =
      String.to_charlist(s)
      |> Enum.take(mid)

    back =
      String.to_charlist(s)
      |> Enum.reverse()
      |> Enum.take(mid)

    front == back
  end
end

defmodule TestValidPalindrome do
  use ExUnit.Case

  test "leetcode examples" do
    assert ValidPalindrome.valid?("racecar") == true
    assert ValidPalindrome.valid?("raceacar") == false
    assert ValidPalindrome.valid?(" ") == true
  end
end
